### PR TITLE
Feature/openapi templates

### PR DESCRIPTION
**Pull request by bretterer**
Builds the templates for Openapi Generator

This is for the initial release of the generator templates.   More changes and additions will be included in a later PR along with the tests.

To build, run `make build`  

Clean up files that are generated, run `make clean-files`

To regenerate only the files without cleaning first, run `make generate-files`

Pull down a new spec, `make pull-spec`

### CODE REVIEW

File: vendor/github.com/go-yaml/yaml/resolve.go [hunk 1]
```diff
@@ -180,7 +180,7 @@ func resolve(tag string, in string) (rtag string, out interface{}) {
 					return yaml_INT_TAG, uintv
 				}
 			} else if strings.HasPrefix(plain, "-0b") {
-				intv, err := strconv.ParseInt("-" + plain[3:], 2, 64)
+				intv, err := strconv.ParseInt("-"+plain[3:], 2, 64)
```
Comment (bdemers) [inline vendor/github.com/go-yaml/yaml/resolve.go]: Another nit, some string concats have spaces, others don't

Comment (bretterer) [reply vendor/github.com/go-yaml/yaml/resolve.go]: This is within the vendor file... no control over it.... Guessing when I ran `go-fmt` it pulled in the stuff from vendor

File: okta/requestExecutor.go [hunk 2]
```diff
@@ -59,25 +59,25 @@ func (re *RequestExecutor) doRequest(method string, url string, body io.Reader)
...
+	if err != nil {
 		return nil, err
 	}
 
-	if (method == "POST" && body != nil) {
+	if method == "POST" && body != nil {
 		req.Header.Add("Content-Type", "application/json")
 	}
-	req.Header.Add("Authorization", "SSWS " + re.config.Okta.Client.Token)
+	req.Header.Add("Authorization", "SSWS "+re.config.Okta.Client.Token)
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/requestExecutor.go]: `go-fmt` formats it this way.

File: okta/userFactor.go [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+type Factor struct {
+	Embedded map[string]interface{} `json:"_embedded, omitempty"`
+	Links map[string]interface{} `json:"_links, omitempty"`
+	Device string `json:"device, omitempty"`
+	DeviceType string `json:"deviceType, omitempty"`
+	FactorType string `json:"factorType, omitempty"`
+	Id string `json:"id, omitempty"`
+	MfaStateTokenId string `json:"mfaStateTokenId, omitempty"`
+	Profile FactorProfile `json:"profile, omitempty"`
+	Provider string `json:"provider, omitempty"`
```
Comment (laura-rodriguez) [inline okta/userFactor.go]: Shouldn't be a type called `FactorProvider` for `Provider` defined?

File: okta/userFactor.go [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+	RechallengeExistingFactor bool `json:"rechallengeExistingFactor, omitempty"`
+	SessionId string `json:"sessionId, omitempty"`
+	Status string `json:"status, omitempty"`
```
Comment (laura-rodriguez) [inline okta/userFactor.go]: Missing `FactorStatus` type here?

File: okta/userFactor.go [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+
+	r := VerifyFactorResponse{}
+
+	json.Unmarshal(resp, &r)
+
+	return r
+}
+
+
+type FactorProfile struct {
```
Comment (laura-rodriguez) [inline okta/userFactor.go]: Several empty structs here...

File: okta/application.go [hunk 6]
```diff
@@ -0,0 +1,1267 @@
...
+
+import (
+	"bytes"
+	"fmt"
+	"encoding/json"
+	"time"
+)
+
+type AppUser struct {
+	Embedded map[string]interface{} `json:"_embedded, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: They are available for the end user.  I do not explicitly use them anywhere

File: okta/application.go [hunk 7]
```diff
@@ -0,0 +1,1267 @@
...
+
+type ApplicationResource resource
+
+type Application struct {
+	Embedded map[string]interface{} `json:"_embedded, omitempty"`
+	Links map[string]interface{} `json:"_links, omitempty"`
+	Accessibility ApplicationAccessibility `json:"accessibility, omitempty"`
+	Created time.Time `json:"created, omitempty"`
+	Credentials ApplicationCredentials `json:"credentials, omitempty"`
+	Features map[string]interface{} `json:"features, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: It is a map according to the spec.

File: okta/application.go [hunk 8]
```diff
@@ -0,0 +1,1267 @@
...
+
+	r := AppUser{}
+
+	json.Unmarshal(resp, &r)
+
+	return r
+}
+
+func (m *ApplicationResource) Getapplicationuser(appId string, userId string) AppUser {
+	resp, err := m.client.requestExecutor.Get("/api/v1/apps/"+appId+"/users/"+userId+"")
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: Auto generated code... it is the same across all of them and sometimes, things are added there.

File: okta/application.go [hunk 9]
```diff
@@ -0,0 +1,1267 @@
...
+}
+
+
+type ApplicationGroupAssignment struct {
+	Embedded map[string]interface{} `json:"_embedded, omitempty"`
+	Links map[string]interface{} `json:"_links, omitempty"`
+	Id string `json:"id, omitempty"`
+	LastUpdated time.Time `json:"lastUpdated, omitempty"`
+	Priority int64 `json:"priority, omitempty"`
+	Profile map[string]interface{} `json:"profile, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: Just making it available to use as needed... profile is in the spec, so it made it to the SDK

File: okta/application.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: Never mind.. you are right... I will put in a change

File: okta/application.go [hunk 10]
```diff
@@ -0,0 +1,1267 @@
...
+}
+func (m *ApplicationSettingsNotificationsVpn) WithNetwork(v ApplicationSettingsNotificationsVpnNetwork) *ApplicationSettingsNotificationsVpn {
+	m.Network = v
+	return m
+}
+
+
+type ApplicationSettingsNotificationsVpnNetwork struct {
+	Connection string `json:"connection, omitempty"`
+	Exclude map[string]interface{} `json:"exclude, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: Yup... Ill change this

File: okta/application.go [hunk 11]
```diff
@@ -0,0 +1,1267 @@
...
+	m.Exclude = v
+	return m
+}
+func (m *ApplicationSettingsNotificationsVpnNetwork) WithInclude(v map[string]interface{}) *ApplicationSettingsNotificationsVpnNetwork {
+	m.Include = v
+	return m
+}
+
+
+type ApplicationSignOnMode struct {
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: no enums in golang

File: okta/application.go [hunk 12]
```diff
@@ -0,0 +1,1267 @@
...
+	m.ResponseTypes = v
+	return m
+}
+func (m *OpenIdConnectApplicationSettingsClient) WithTosUri(v string) *OpenIdConnectApplicationSettingsClient {
+	m.TosUri = v
+	return m
+}
+
+
+type OpenIdConnectApplicationType struct {
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: No enums in Golang

File: okta/application.go [hunk 13]
```diff
@@ -0,0 +1,1267 @@
...
+	AssertionSigned bool `json:"assertionSigned, omitempty"`
+	AttributeStatements map[string]interface{} `json:"attributeStatements, omitempty"`
+	Audience string `json:"audience, omitempty"`
+	AudienceOverride string `json:"audienceOverride, omitempty"`
+	AuthnContextClassRef string `json:"authnContextClassRef, omitempty"`
+	DefaultRelayState string `json:"defaultRelayState, omitempty"`
+	Destination string `json:"destination, omitempty"`
+	DestinationOverride string `json:"destinationOverride, omitempty"`
+	DigestAlgorithm string `json:"digestAlgorithm, omitempty"`
+	HonorForceAuthn bool `json:"honorForceAuthn, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: This is something ill think about.   I like that the property name matches what our docs state

File: okta/application.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: Yup

File: okta/session.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+type SessionIdentityProvider struct {
+	Id string `json:"id, omitempty"`
+	Type string `json:"type, omitempty"`
+}
+
+
+
+type SessionIdentityProviderType struct {
+}
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/session.go]: No Enum

File: okta/session.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+
+
+type SessionStatus struct {
+}
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/session.go]: No Enum

File: okta/userFactor.go (hunk 3 above)
Comment (bretterer) [reply okta/userFactor.go]: object model is undefined for provider in the spec

File: okta/userFactor.go (hunk 4 above)
Comment (bretterer) [reply okta/userFactor.go]: object model is undefined in the spec

File: okta/userFactor.go (hunk 5 above)
Comment (bretterer) [reply okta/userFactor.go]: No Enums in Golang

File: okta/application.go [hunk 16]
```diff
@@ -0,0 +1,1267 @@
...
+func (m *SamlAttributeStatement) WithValues(v map[string]interface{}) *SamlAttributeStatement {
+	m.Values = v
+	return m
+}
+
+
+type SchemeApplicationCredentials struct {
+	Password PasswordCredential `json:"password, omitempty"`
+	RevealPassword bool `json:"revealPassword, omitempty"`
+	Scheme string `json:"scheme, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: object model undefined

File: okta/application.go [hunk 17]
```diff
@@ -0,0 +1,1267 @@
...
+func (m *WsFederationApplicationSettings) WithApp(v WsFederationApplicationSettingsApplication) *WsFederationApplicationSettings {
+	m.App = v
+	return m
+}
+
+
+type WsFederationApplicationSettingsApplication struct {
+	AttributeStatements string `json:"attributeStatements, omitempty"`
+	AudienceRestriction string `json:"audienceRestriction, omitempty"`
+	AuthnContextClassRef string `json:"authnContextClassRef, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/application.go]: see previous comment about renaming properties

File: okta/group.go [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func (m *GroupResource) Removeuser(groupId string, userId string)  {
+	_, err := m.client.requestExecutor.Delete("/api/v1/groups/"+groupId+"/users/"+userId+"")
+	if err != nil  {
+		fmt.Printf("error: %v", err)
+	}
+}
+
+func (m *GroupResource) Listusers(groupId string) User {
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/group.go]: old codegen... generated models are not supplied in this anymore.   it is now ListUsers

File: okta/log.go [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+	CredentialProvider map[string]interface{} `json:"credentialProvider, omitempty"`
+	CredentialType map[string]interface{} `json:"credentialType, omitempty"`
+	ExternalSessionId string `json:"externalSessionId, omitempty"`
+	Interface string `json:"interface, omitempty"`
+	Issuer LogIssuer `json:"issuer, omitempty"`
+}
+
+
+
+type LogAuthenticationProvider struct {
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/log.go]: No Enum

File: okta/log.go [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+	Id string `json:"id, omitempty"`
+	IpAddress string `json:"ipAddress, omitempty"`
+	UserAgent LogUserAgent `json:"userAgent, omitempty"`
+	Zone string `json:"zone, omitempty"`
+}
+
+
+
+type LogCredentialProvider struct {
+}
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/log.go]: No Enum

File: okta/log.go [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+
+
+
+type LogCredentialType struct {
+}
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/log.go]: No Enum

File: okta/log.go [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+	Country string `json:"country, omitempty"`
+	Geolocation LogGeolocation `json:"geolocation, omitempty"`
+	PostalCode string `json:"postalCode, omitempty"`
+	State string `json:"state, omitempty"`
+}
+
+
+
+type LogGeolocation struct {
+	Lat float64 `json:"lat, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/log.go]: See previous comment about renaming

File: okta/log.go [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+type LogOutcome struct {
+	Reason string `json:"reason, omitempty"`
+	Result string `json:"result, omitempty"`
+}
+
+
+
+type LogRequest struct {
+	IpChain map[string]interface{} `json:"ipChain, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/log.go]: Spec seems to not report object model correctly

File: okta/log.go [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+	AsOrg string `json:"asOrg, omitempty"`
+	Domain string `json:"domain, omitempty"`
+	IsProxy bool `json:"isProxy, omitempty"`
+	Isp string `json:"isp, omitempty"`
+}
+
+
+
+type LogSeverity struct {
+}
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/log.go]: No Enum

File: okta/log.go [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+	Detail map[string]interface{} `json:"detail, omitempty"`
+	Id string `json:"id, omitempty"`
+	Type string `json:"type, omitempty"`
+}
+
+
+
+type LogUserAgent struct {
+	Browser string `json:"browser, omitempty"`
+	Os string `json:"os, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/log.go]: See renaming comment

File: okta/session.go [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+func (m *CreateSessionRequest) WithSessionToken(v string) *CreateSessionRequest {
+	m.SessionToken = v
+	return m
+}
+
+type SessionResource resource
+
+type Session struct {
+	Links map[string]interface{} `json:"_links, omitempty"`
+	Amr map[string]interface{} `json:"amr, omitempty"`
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/session.go]: See Renaming Comment

File: okta/session.go [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+	r := Session{}
+
+	json.Unmarshal(resp, &r)
+
+	return r
+}
+
+
+type SessionAuthenticationMethod struct {
+}
```
[the remark this answers is not in the document]
Comment (bretterer) [reply okta/session.go]: No Enum

[the remark this answers is not in the document]
Comment (laura-rodriguez) [reply okta/session.go]: so, what does empty struct mean?

File: vendor/github.com/go-yaml/yaml/resolve.go (hunk 1 above)
Comment (bdemers) [reply vendor/github.com/go-yaml/yaml/resolve.go]: oh, sorry, i didn't notice the `/vendor` path

**Pull request merged**